internal/credspropagation/openstack.go

```diff
@@ -0,0 +1,273 @@
+func fetchOpenStackSecret(ctx context.Context, cfg *PropagationCfg, credential *hmc.Credential) (*corev1.Secret, error) {
+	if credential.Spec.IdentityRef == nil {
+		return nil, errors.New("credential.Spec.IdentityRef is nil")
+	}
+
+	openstackSecret := &corev1.Secret{}
+	if err := cfg.Client.Get(ctx, client.ObjectKey{
+		Name:      credential.Spec.IdentityRef.Name,
+		Namespace: credential.Spec.IdentityRef.Namespace,
+	}, openstackSecret); err != nil {
+		return nil, fmt.Errorf("failed to get OpenStack secret %s/%s: %w",
+			credential.Spec.IdentityRef.Namespace, credential.Spec.IdentityRef.Name, err)
+	}
+	return openstackSecret, nil
+}
+
+// Generate the CCM secret from the OpenStack secret
+func generateOpenStackCCMSecret(ctx context.Context, cfg *PropagationCfg, openstackSecret *corev1.Secret, cloudName string) (*corev1.Secret, error) {
+	const cloudConfTemplate = `
+[Global]
+auth-url="{{ .AuthURL }}"
+{{- if .ApplicationCredentialID }}
+application-credential-id="{{ .ApplicationCredentialID }}"
+{{- end }}
+{{- if .ApplicationCredentialName }}
+application-credential-name="{{ .ApplicationCredentialName }}"
+{{- end }}
+{{- if .ApplicationCredentialSecret }}
+application-credential-secret="{{ .ApplicationCredentialSecret }}"
+{{- end }}
+{{- if and (not .ApplicationCredentialID) (not .ApplicationCredentialSecret) }}
+username="{{ .Username }}"
+password="{{ .Password }}"
+{{- end }}
+region="{{ .RegionName }}"
+interface="{{ .Interface }}"
+
+[LoadBalancer]
+{{- if .FloatingNetworkID }}
+floating-network-id="{{ .FloatingNetworkID }}"
+{{- end }}
+
+[Network]
+{{- if .PublicNetworkName }}
+public-network-name="{{ .PublicNetworkName }}"
+{{- end }}
+`
+
+	// Parse the clouds.yaml content
+	cloudsYaml, ok := openstackSecret.Data["clouds.yaml"]
+	if !ok {
+		return nil, errors.New("missing clouds.yaml in OpenStack secret")
+	}
+
+	parsedCloudsYaml, err := parseCloudsYaml(cloudsYaml)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse clouds.yaml: %w", err)
+	}
+
+	// Extract CloudConfFields using the provided cloudName
+	fields, err := extractCloudConfFields(parsedCloudsYaml, cloudName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract cloud.conf fields: %w", err)
+	}
+
+	// Fetch external network details from OpenStackCluster
+	externalNetwork, err := fetchExternalNetwork(ctx, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch external network details: %w", err)
+	}
+	if externalNetwork == nil {
+		return nil, errors.New("external network is nil")
+	}
+	if externalNetwork.ID == "" || externalNetwork.Name == "" {
+		return nil, errors.New("external network details not found")
+	}
+	fields.FloatingNetworkID = externalNetwork.ID
+	fields.PublicNetworkName = externalNetwork.Name
+
+	// Render the cloud.conf secret
+	return renderCloudConf(cloudConfTemplate, fields)
+}
+
+// Parse the clouds.yaml content into structured types
+func parseCloudsYaml(data []byte) (*CloudsYaml, error) {
+	var cloudsYaml CloudsYaml
```

Comment:
```suggestion
	var cy cloudsYaml
```

Avoid shadowing (lint after the unexporting rename)